src/lib.rs

```diff
@@ -1,101 +1,61 @@
 
-use anyhow::{bail, Context, Result};
-use ldap_poller::{ldap::EntryStatus, ldap3::SearchEntry, Cache, Ldap};
-use tokio::sync::mpsc::Receiver;
-use tokio_stream::{wrappers::ReceiverStream, StreamExt};
+use anyhow::Result;
 
 mod config;
+mod source_ldap;
+mod source_list;
 mod user;
 mod zitadel;
 
 pub use config::{AttributeMapping, Config, FeatureFlag};
+use source_ldap::SourceLdap;
+use source_list::SourceList;
+use user::User;
 use zitadel::Zitadel;
 
-/// Run the sync
-pub async fn sync_ldap_users_to_zitadel(config: Config) -> Result<()> {
-	if !config.feature_flags.contains(&FeatureFlag::SsoLogin) {
-		anyhow::bail!("Non-SSO configuration is currently not supported");
-	}
-
-	let cache = read_cache(&config.cache_path).await?;
-	let zitadel = Zitadel::new(&config).await?;
-	let (mut ldap_client, ldap_receiver) = Ldap::new(config.clone().ldap.into(), cache);
-
-	let deactivate_only = config.deactivate_only();
+/// Trait to define sources that can be used for syncing
+trait Source {
+	/// Create a new sync source
+	fn new(config: &Config) -> Result<Self>
+	where
+		Self: Sized;
 
-	let sync_handle: tokio::task::JoinHandle<Result<_>> = tokio::spawn(async move {
-		ldap_client.sync_once(None).await.context("failed to sync/fetch data from LDAP")?;
+	/// Get lists of all changes
+	async fn get_all_changes(&self) -> Result<(Vec<User>, Vec<(User, User)>, Vec<String>)>;
 
-		if config.dry_run() {
-			tracing::warn!("Not writing ldap cache during a dry run");
-		} else {
-			let cache = ldap_client.persist_cache().await;
-			tokio::fs::write(
-				&config.cache_path,
-				bincode::serialize(&cache).context("failed to serialize cache")?,
-			)
-			.await
-			.context("failed to write cache")?;
-		}
+	/// Get list of user emails that have been removed
+	async fn get_removed_user_emails(&self) -> Result<Vec<String>>;
+}
 
-		tracing::info!("Finished syncing LDAP data");
+/// Perform a sync operation
+pub async fn perform_sync(config: &Config) -> Result<()> {
+	if !config.feature_flags.require_sso_login() {
+		anyhow::bail!("Non-SSO configuration is currently not supported");
+	}
 
-		Ok(())
-	});
+	// Setup Zitadel client
+	let zitadel = Zitadel::new(config).await?;
 
-	let (added, changed, removed) = get_user_changes(ldap_receiver).await;
+	// Perform LDAP sync
+	if config.source_ldap.is_some() {
+		let ldap_sync = SourceLdap::new(config)?;
+		let (added, changed, removed) = ldap_sync.get_all_changes().await?;
 
-	sync_handle.await??;
+		if !config.feature_flags.deactivate_only() {
+			zitadel.import_new_users(added).await?;
+			zitadel.delete_users_by_id(removed).await?;
+		}
 
-	zitadel.update_users(changed).await?;
+		zitadel.update_users(changed).await?;
+	}
 
-	if !deactivate_only {
-		zitadel.import_new_users(added).await?;
-		zitadel.delete_users(removed).await?;
+	// Perform Disable List sync
+	if config.source_list.is_some() {
+		let endpoint_sync = SourceList::new(config)?;
+		let removed = endpoint_sync.get_removed_user_emails().await?;
```

Comment:
Can't we just also use `.get_all_changes` here, except only fill out the email? Having a trait with two methods, one of which is implemented for one implementation, and the other for the other, but neither is shared, just sounds like we shouldn't be using a trait.

Reply:
Good point, will do that.